util/gitutil/git_ref_test.go

```diff
@@ -0,0 +1,26 @@
+package gitutil
+
+import "testing"
+
+func TestIsGitRef(t *testing.T) {
+	cases := map[string]bool{
+		"https://example.com/":                 false,
+		"https://example.com/foo":              false,
+		"https://example.com/foo.git":          true,
+		"https://example.com/foo.git#deadbeef": true,
+		"https://example.com/foo.git/":         false,
+		"https://example.com/foo.git.bar":      false,
+		"git://example.com/foo":                true,
```

Comment:
I'm not sure.
Unencrypted connections might be acceptable for public repos when we validate the commit hash (#2816) ?

Reply:
My train of thought was that;

- `git://` has been misinterpreted by many to be "secure" (where it is not); I think this was one of the reasons for GitHub deprecating it
- as with anything; it's easier to _add_ support (if there's strong demand) than to _remove_

Do we know how _common_ it is to still use? (Instead of `http://`  (without authentication?) for read-only)? `http://` is  slightly clearer that it's un-encrypted.

Reply:
On the opposite side; wondering if `file://<path in context>.git` could allow interesting use-cases, e.g. to be able to add only files in version control; as a (somewhat) alternative to `.dockerignore` etc 🤔

Reply:
Updated PR to reject `git://` and `http://` for `ADD` instructions.

```go
if gitRef.UnencryptedTCP {
  return errors.New("unencrypted TCP is not supported for ADD <git ref>")
}
```

Reply:
> file://<path in context>.git

I'm not sure. Some users may except this form to accept an arbitrary file path in the client (which won't be implemented).

Reply:
Why should we do anything different for this that what is supported by the git build context? For the `file://` I don't really see a use case atm. If the files are already in build context then just use them directly.

`git://` and `http://` are not the same protocol afaik. As long as `git` supports them I don't see a reason for us to block them. If we want to move to `go-git` implementation in the future we might need to drop dump-http but that seems unrelated.

Reply:
Updated the PR to allow UnencryptedTCP